src/index.ts

```diff
@@ -90,22 +93,32 @@ export default class Kumulos {
         }
     }
 
-    getInstallId(): Promise<InstallId> {
-        return getInstallId();
-    }
+    private async maybePersistInstallIdAndUserId(config: KumulosConfig) : Promise<void> {
+        await getInstallId()
```

Comment:
Do we need to get then set? Could this cause race conditions with the initial installation events?

I think it looks like we may need to just always set the install ID and defer continuation until completed?

Reply:
(OK race perhaps not due to the promise, but if we generated the UUID and resolved get with it, then we'd never see the newly set one?)

Reply:
why would we never see the new one? it updates every time it's not equal to the one we give here

Reply:
Hm I _think_ perhaps because we do something like:

setInstallId (async)
main init then continues with track install details, which will read install ID
this creates one promise which resolves & caches ID
if ID not set, it's generated
both ops async

any potential to race / have invalid state?

Maybe because operations would be queued in the event loop no issue, but just seems a little bit off to me, let's discuss in more detail tomorrow.

Reply:
lol, at least the intention was to wait until install id is set. prob it indeed doesnt wait. Return the promise given by setInstallId enough?